cuda/matrix/hybrid_kernels.cu

```diff
@@ -0,0 +1,436 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/hybrid_kernels.hpp"
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/matrix/ell.hpp>
+
+
+#include "core/matrix/coo_kernels.hpp"
+#include "core/matrix/ell_kernels.hpp"
+#include "cuda/base/types.hpp"
+#include "cuda/components/atomic.cuh"
+#include "cuda/components/cooperative_groups.cuh"
+#include "cuda/components/prefix_sum.cuh"
+#include "cuda/components/reduction.cuh"
+#include "cuda/components/zero_array.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace cuda {
+/**
+ * @brief The Hybrid matrix format namespace.
+ *
+ * @ingroup hybrid
+ */
+namespace hybrid {
+
+
+constexpr int default_block_size = 512;
+constexpr int warps_in_block = 4;
+
+
+template <typename ValueType, typename IndexType>
+void convert_to_dense(
+    std::shared_ptr<const CudaExecutor> exec, matrix::Dense<ValueType> *result,
+    const matrix::Hybrid<ValueType, IndexType> *source) GKO_NOT_IMPLEMENTED;
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_HYBRID_CONVERT_TO_DENSE_KERNEL);
+
+
+namespace kernel {
+
+
+// Copied from coo_kernel.cu
+template <int subwarp_size = cuda_config::warp_size, typename ValueType,
+          typename IndexType>
+__device__ __forceinline__ void segment_scan(
+    const group::thread_block_tile<subwarp_size> &group, IndexType ind,
+    ValueType *__restrict__ val, bool *__restrict__ head)
+{
+#pragma unroll
+    for (int i = 1; i < subwarp_size; i <<= 1) {
+        const IndexType add_ind = group.shfl_up(ind, i);
+        ValueType add_val = zero<ValueType>();
+        if (threadIdx.x >= i && add_ind == ind) {
+            add_val = *val;
+            if (i == 1) {
+                *head = false;
+            }
+        }
+        add_val = group.shfl_down(add_val, i);
+        if (threadIdx.x < subwarp_size - i) {
+            *val += add_val;
+        }
+    }
+}
+
+
+/**
+ * The global function for counting nonzeros of COO.
+ * It is almost like COO spmv routine.
+ *
+ * @param nnz  the number of nonzeros in the matrix
+ * @param num_line  the maximum round of each warp
+ * @param val  the value array of the matrix
+ * @param row  the row index array of the matrix
+ * @param c  the output nonzeros per row
+ */
```

Comment:
Please document the template parameters. In particular, what is SizeType supposed to be (and why is it a template parameter?).